apm-agent-plugins/apm-es-restclient-plugin/src/main/java/co/elastic/apm/es/restclient/ESRestClientInstrumentationHelper.java

```diff
@@ -0,0 +1,59 @@
+/*-
+ * #%L
+ * Elastic APM Java agent
+ * %%
+ * Copyright (C) 2018 Elastic and contributors
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+package co.elastic.apm.es.restclient;
+
+import co.elastic.apm.bci.VisibleForAdvice;
+import co.elastic.apm.report.serialize.DslJsonSerializer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.io.InputStream;
+
+@VisibleForAdvice
+public class ESRestClientInstrumentationHelper {
+    private static final Logger LOGGER = LoggerFactory.getLogger(ESRestClientInstrumentationHelper.class);
+
+    private static ThreadLocal<byte[]> bodyReadBuffer = new ThreadLocal<>();
+
+    @Nullable
+    @VisibleForAdvice
+    public static String readRequestBody(InputStream bodyIS, String endpoint) throws IOException {
+        String body = null;
+        try {
+            byte[] data = bodyReadBuffer.get();
+            if (data == null) {
+                bodyReadBuffer.set(new byte[DslJsonSerializer.MAX_LONG_STRING_VALUE_LENGTH]);
+                data = bodyReadBuffer.get();
+            }
+            int length = bodyIS.read(data, 0, data.length);
+
+            // Using platform's default charset for decoding
+            body = new String(data, 0, length);
```

Comment:
One idea how we might be able to avoid allocations altogether is to use a `CharBuffer statementBuffer` field in `co.elastic.apm.impl.transaction.Db`. Together with [`CharsetDecoder#decode`](https://docs.oracle.com/javase/7/docs/api/java/nio/charset/CharsetDecoder.html#decode(java.nio.ByteBuffer,%20java.nio.CharBuffer,%20boolean)) and your thread-local byte array buffer, we would never create garbage as we can recycle the `CharBuffer` via the `rewind` method as well as the thread-local ByteBuffer. `CharBuffer` also implements the `CharSequence` interface which DSL JSON can directly serialize.

Reply:
If you prefer keeping the first version simple, we can also revisit that idea later, after merging the PR. This is just meant as an idea 🙂

Reply:
That's a good idea. Let's talk about what to do with locating the proper charset and encoding/decoding because I think that's a bigger issue

Reply:
Looked into doing that, but Db is instantiated per Span, and there can be **A LOT** of those, potentially most or all of which are not DB span. This may lead to big allocation of constant memory, which can be worse than reallocation of recyclable memory. If we really want to do that properly, we need to pool those separately. 
Leaving it as it is at the moment.